Contracts nightly

EricWF opened the pull request.

EricWF committed:
Add contracts representation and parsing. Much thanks to Peter for the initial implementation

EricWF committed:
Add contracts representation and parsing. Much thanks to Peter for the initial implementation

EricWF committed:
fix todo

EricWF committed:
fix todo

EricWF committed:
get it working end-to-end

EricWF committed:
get it working end-to-end

EricWF committed:
a bunch of work, most of it broken

EricWF committed:
a bunch of work, most of it broken

EricWF committed:
Get templates working.

The template stuff is a bit of a hack. I'm considering simply inserting
the statements into the function body and letting the existing code
handle the rest.

Also add some cheaky diagnostics, fix a few name lookup bugs, and
delete some old code.

EricWF committed:
Get templates working.

The template stuff is a bit of a hack. I'm considering simply inserting
the statements into the function body and letting the existing code
handle the rest.

Also add some cheaky diagnostics, fix a few name lookup bugs, and
delete some old code.

EricWF committed:
Improve codegen, Start on constification.

There are also various experiements that need cleaning up,
specifically around builtins and source location.

EricWF committed:
Improve codegen, Start on constification.

There are also various experiements that need cleaning up,
specifically around builtins and source location.

EricWF committed:
A lot of forward progress.

Some work on constification, but a bunch of experiment that still needs
to be undone.

The addition of -fclang-contract-groups=foo=<semantic>, which still
needs better integrating with -fcontract-evaluation-semantic=.

The [[clang::contract_group("foo.bar")]] stuff works really well.
I'll be happy to try it in libc++.

What still needs to be done includes:
  * Coming up with a proper descriptor/format for the compiler-stl
    interface. It seems to me we'll want the ABI to be able to evolve
over time, and so we should pass a pointer to data + data descriptor of
some kind so that the STL can parse version of the header from different
compilers & versions.

  * Everything to do with parsing pre/post in inline member functions.
  * Contract redeclaration checking.
  * Constifying implicit this.
  * Code gen & constant evaluation for result names (and representation
    too).
  * Cleanup around experiments for source location & constification.

EricWF committed:
A lot of forward progress.

Some work on constification, but a bunch of experiment that still needs
to be undone.

The addition of -fclang-contract-groups=foo=<semantic>, which still
needs better integrating with -fcontract-evaluation-semantic=.

The [[clang::contract_group("foo.bar")]] stuff works really well.
I'll be happy to try it in libc++.

What still needs to be done includes:
  * Coming up with a proper descriptor/format for the compiler-stl
    interface. It seems to me we'll want the ABI to be able to evolve
over time, and so we should pass a pointer to data + data descriptor of
some kind so that the STL can parse version of the header from different
compilers & versions.

  * Everything to do with parsing pre/post in inline member functions.
  * Contract redeclaration checking.
  * Constifying implicit this.
  * Code gen & constant evaluation for result names (and representation
    too).
  * Cleanup around experiments for source location & constification.

EricWF committed:
whitespace changes

EricWF committed:
whitespace changes

EricWF committed:
change flags, but break clang

EricWF committed:
change flags, but break clang

EricWF committed:
finish merge

EricWF committed:
get constification working for "this"

EricWF committed:
Fix constexpr eval and bug in contract group

EricWF committed:
Fix codegen for trap

EricWF committed:
Emit the contract assertion info as a single struct

This allows better versioning, and also prevents us from clobbering
every registery. (though that latter case is likely not too impactful).

EricWF committed:
Get codegen working with exceptions

EricWF committed:
Fix cleanups by finishing statement.

Also, re-enable the exception handling tentatively.

EricWF committed:
Fix some modules bugs & fix exception violation handling

EricWF committed:
Temp work save

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts-reset

EricWF committed:
remove temp asserts

EricWF committed:
rename ContractOpts to prevent GCC from getting mad

EricWF committed:
Fix build with GCC but introduce issues in constant evaluation

EricWF committed:
replace assertion with TODO diagnostic

EricWF committed:
fixup more tests

EricWF committed:
not working, but progress

EricWF committed:
Constexpr evaluation hacked together, don't really understand it all yet

EricWF committed:
Got constant expressions working.

Now just a ton of cleanup to go.

EricWF committed:
Merge branch 'contracts-main' of github.com:efcs/llvm-project into eric-contracts

EricWF committed:
start expr constant cleanup

EricWF committed:
Finish constexpr cleanup - candidate for nightly

EricWF committed:
remove duplicate test

EricWF committed:
It's a bunch of progress, but a big mess.

Member functions are coming together,
I've formalized the building and type of the contract violation as
passed.
I've made codegen pass globals rather that creating the object on the
fly. The IR is a lot cleaner, we'll see if I stick with that.

EricWF committed:
Fixes for nightly build

EricWF committed:
Merge branch 'main' into eric-contracts

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts

EricWF committed:
Get post codegen working.

EricWF committed:
Add more tests, fix bug in constant evaluation

EricWF committed:
Get contract codegen working with phi dispatching.

This cleans up the call graph and emitted code a bunch.
Still a bunch of cleanup to do in the actual CodeGenFunction code though

EricWF committed:
remove bad comment, start libc++ tests

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts

EricWF committed:
resolve parsing regression

EricWF committed:
Add some new kinds of tests, also some junk temp cmake, get post working better at code gen time

EricWF committed:
cleanup more code

EricWF committed:
Save progress.

This commit is in the middle of a bunch of changes, in particular to
codegen. However, it's also at a state where everything works, and I
haven't committed in a while, so this commit (while a mess) is to save
progress.

EricWF committed:
Merge branch 'main' into eric-contracts

EricWF committed:
Save libc++ improvements

EricWF committed:
Merge branch 'main' into eric-contracts

EricWF committed:
Cleanup, Fixes, and restart of codegen.

There's a bunch of cleanup for parsing here. Including fixing a couple
of bugs where we try too eagerly to form a return type for parsing,
causing entirely unrelated code to fail to parse.

Also, changes to the resultNameDecl representation are coming.
The intent of these changes is to create a "canonical" decl which
can be used by _all post conditions_ on the function to use as a
key, allowing codegen & constant evaluation to all reference the
same lvalue.

EricWF committed:
Messy Save.

Everything is passing, no regressions, but things are a bit messy.

Still, it's been a while since I pushed a nightly.

EricWF committed:
Remove debug code.

EricWF committed:
merge

EricWF committed:
Merge remote-tracking branch 'upstream/main' into eric-contracts

EricWF committed:
Add some tests for contracts.

EricWF committed:
rework codegen, not working, but need to merge

EricWF committed:
merge

EricWF committed:
Rework codegen so that exceptions are hopefully supported...

EricWF committed:
Fix codegen for throwing contracts

EricWF committed:
Some test cleanup

EricWF committed:
get more working

EricWF committed:
Get constant initialization working.

EricWF committed:
Fix bug in constant evaluation reusing previously computed value

EricWF committed:
Add diagnostics for redeclarations, but redeclarations dont quite work
yet.

EricWF committed:
Improvements, and some broken expr constant stuff

EricWF committed:
mostly working

EricWF committed:
Fix GCC build hopefully

EricWF committed:
working save. needs cleanup

EricWF committed:
remove a bunch of debug print lines

EricWF committed:
suppress warning

EricWF committed:
remove and replace bad assertions

EricWF committed:
cleanup for deduced auto

EricWF committed:
Lazy parsing and support for trailing auto

EricWF committed:
try adding builders

EricWF committed:
adjust glob

EricWF committed:
get working

EricWF committed:
checkout sources first

EricWF committed:
fix build

EricWF committed:
a bunch of dummy code and cleanup

EricWF committed:
almost clean test suite

EricWF committed:
get tests passing

EricWF committed:
fixup

EricWF committed:
changes

EricWF committed:
all changes

EricWF committed:
all changes

EricWF committed:
try to add verification job

EricWF committed:
try to add verification job

EricWF committed:
force ci

EricWF committed:
remove garbage file

EricWF committed:
Cleanup headers

EricWF committed:
make almost everything worse

EricWF committed:
remove garbage

EricWF committed:
merge with upstream

EricWF committed:
Cleanup and format

EricWF committed:
Start work on virtual functions by adding caller side emmission.

This is kind of two birds one stone, because it may be better to use
funclets or some such yet-to-be-understood thing.

EricWF committed:
Merge branch 'main' into eric-contracts

EricWF committed:
start representation overhall

EricWF committed:
Before merge

EricWF committed:
attempt bad merge

EricWF committed:
Start a very large refactor of template machinary and sema structure.

EricWF committed:
A working mess.

After a week of refactoring and overhalls, I have everything working,
now I just need to clean it up.

Well everything except debug info.

EricWF committed:
remove dead code

EricWF committed:
fix up tests, get templates working better

EricWF committed:
Add more tests

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts

EricWF committed:
more cleanup

EricWF committed:
diagnostic improvements

EricWF committed:
Fix a number of crashes

EricWF committed:
merge

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts

EricWF committed:
everything except libcxx working

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into eric-contracts

EricWF committed:
I think I got lambdas working-ish

EricWF committed:
remove print code

EricWF committed:
remove xfails, is this working?

EricWF committed:
start work on instantiation stuff

EricWF committed:
adjust libcxx tests

EricWF committed:
get contract diagnosing almost working

EricWF committed:
save progress

EricWF committed:
save before removing old code

EricWF committed:
remove a bunch of bad code

EricWF committed:
get working

EricWF committed:
More cleanup, add crashing lambda tests

EricWF committed:
various saves

EricWF committed:
merge with upstream

EricWF committed:
save state

EricWF committed:
with working tests

EricWF committed:
broken

EricWF committed:
idk

EricWF committed:
llvm fixes v2

EricWF committed:
delete

EricWF committed:
merge

EricWF committed:
address equality bug

EricWF committed:
add test

EricWF committed:
merge

EricWF committed:
finish merge

EricWF committed:
merge

EricWF committed:
Merge branch 'main' of github.com:llvm/llvm-project into save-contract-work

EricWF committed:
fix merge issue

EricWF committed:
cleanup and merge fixes

EricWF committed:
add more printouts

EricWF committed:
merge with main

EricWF committed:
complete merge

EricWF committed:
save

EricWF committed:
merge

EricWF committed:
fix merge

EricWF committed:
adjust some failing libc++ tests

EricWF committed:
merge

EricWF committed:
fix build error only present on GCC?

EricWF committed:
finish merge

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Fix build error under GCC

EricWF committed:
Merge w/ Parser changes

EricWF committed:
Add non-cassert parts of D3290R3.

The tests haven't been written yet, since I need to perform a merge
before they'll build (silly upstream libc++, breaking the test suite for
unknown reasons)

EricWF committed:
Merge with contracts-base

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
fix broken build due to new warning

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
add new interfaces

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
fix build failures

EricWF committed:
merge done

EricWF committed:
scope flag rewrites

EricWF committed:
start mangling changes

EricWF committed:
merge with main

EricWF committed:
fix bad merge

EricWF committed:
fix warnings turned into errors

EricWF committed:
add new source location builtin

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
work around C build failures

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
finish merge but broken libcxx tests

EricWF committed:
Fix up post-merge breakages

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
Delete change_default_route_ip.py

EricWF committed:
Delete change_network_config.yml

cor3ntin commented on clang/include/clang/Basic/TokenKinds.def:
```diff
@@ -431,6 +435,10 @@ CXX20_KEYWORD(constinit             , 0)
 
+// C++ contracts keywords
+CONTRACTS_KEYWORD(contract_assert)
+ALIAS("__contract_assert"        , contract_assert      , KEYCXX11)
```
Do we want to bother? We pickerd a terrible keyword because it was not used, and if you have a macro called `__contract_assert`, i sort of don't care

cor3ntin commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1627,6 +1627,33 @@ defm auto_import : BoolFOption<"auto-import",
 def offload_EQ : CommaJoined<["--"], "offload=">, Flags<[NoXarchOption]>, Alias<offload_targets_EQ>,
   HelpText<"Specify comma-separated list of offloading target triples (CUDA and HIP only)">;
 
+// C++ Contracts
+defm contracts : BoolFOption<"contracts",
+  LangOpts<"Contracts">, DefaultFalse,
+  PosFlag<SetTrue, [], [ClangOption, CC1Option, CLOption],
+          "Enable support for C++ Contracts">,
+  NegFlag<SetFalse, [], [ClangOption, CC1Option]>>;
+
```
Cf above, lets forgo that for now

cor3ntin commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1627,6 +1627,33 @@ defm auto_import : BoolFOption<"auto-import",
+  NormalizedValues<["Ignore", "Enforce", "Observe", "QuickEnforce"]>,
+  MarshallingInfoEnum<LangOpts<"ContractOpts.DefaultSemantic">, "Enforce">;
+
+def fcontract_group_evaluation_semantic_EQ : CommaJoined<["-"], "fcontract-group-evaluation-semantic=">,
+  Visibility<[ClangOption, CC1Option]>,
+  HelpText<"Clang extension. Enable or disable contracts by group. The argument is a comma-separated "
+           "sequence of one or more group names, each prefixed by '+' or '-'.">;
+
+defm contract_exceptions : BoolFOption<"contract-exceptions",
+  LangOpts<"ContractExceptions">, DefaultTrue,
+  PosFlag<SetTrue, [], [ClangOption],
+          "Enable contract violation on exception">,
+  NegFlag<SetFalse, [], [ClangOption, CC1Option],
+    "Disable contract violations on exception">>;
```
Can we do that in a subsequent PR?

cor3ntin commented on clang/include/clang/Parse/Parser.h:
```diff
@@ -8933,6 +8936,60 @@ class Parser : public CodeCompletionHandler {
+    LLVM_MARK_AS_BITMASK_ENUM(CES_AllScopes)
+  };
+
+  std::optional<ContractKind> getContractKeyword(const Token &Token) const;
+  std::optional<ContractKind> getContractKeyword() const {
+    return getContractKeyword(Tok);
+  }
+  bool isFunctionContractKeyword() const {
+    return isFunctionContractKeyword(Tok);
+  }
+  bool isFunctionContractKeyword(const Token &Token) const {
+    return getContractKeyword(Token).value_or(ContractKind::Assert) !=
+        ContractKind::Assert;
+  }
+
+  bool isAnyContractKeyword(const Token &Token) const {
+    return getContractKeyword(Token).has_value();
+  }
```
Cf above, lets keep assert separate from pre/post

cor3ntin commented on clang/include/clang/Sema/Scope.h:
```diff
@@ -42,7 +43,7 @@ class Scope {
   /// ScopeFlags - These are bitfields that are or'd together when creating a
   /// scope, which defines the sorts of things the scope contains.
-  enum ScopeFlags {
+  enum ScopeFlags : unsigned long {
```
Maybe we want to introduce an alias for the underlying type

cor3ntin reviewed (commented):
I am far from having done a complete pass, but here are some comments.
Chopping this in manageable pieces will be necessary I'm afraid, Github can barely render the PR and reviewers will be discouraged by the sheer size of it all.

Changes to source location could be a PR and then maybe we want to start with just `contract_assert`, even if i understand it would be annoying for you.

Non essentials (ast matchers for examples) can be separate PRs.
Everything lamba related can also be separated out.

I will keep reviewing today or tomorrow

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+  "woops... FIXME(EricWF): %0"
+>, InGroup<EricWFFixme>;
+
+def err_result_name_shadows_param : Error<
```
Can we be more consistent about using `_contract_` in these diags?

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+def err_result_name_shadows_param : Error<
+  "declaration of result name %0 shadows parameter">;
+def err_void_result_name : Error<
+  "result name %0 cannot be used with a void return type">;
```
used? declared? introduced?

cor3ntin replied:
"result name in the post condition of a function with a void return type" maybe

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+
+
+
+def err_contract_violation : Error<"contract %select{precondition|postcondition|assertion}0 failed%select{: %2|}1">;
+def warn_contract_violation : Warning<"contract %select{precondition|postcondition|assertion}0 failed%select{: %2|}1">,
+    InGroup<ContractViolation>;
+
+def err_contract_requirement_failed : Error<
+  "%select{precondition|postcondition|contract assertion}0 failed due to requirement '%1'%select{: %3|}2">;
+def warn_contract_requirement_failed : Warning<
+  "%select{precondition|postcondition|contract assertion}0 failed due to requirement '%1'%select{: %3|}2">,
+  InGroup<ContractViolation>;
```
None of these are (nor can) ever be used, right?

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+  "parameter %0 declared with type %1 here">;
+
+def err_lambda_implicit_capture_in_contracts_only : Error<
+         "implicit capture of %select{local entity %1|'this'}0 is not allowed when used exclusively in contract assertions">;
```
Can we say instead "must be captured explicitly" ? Bonus point if you add a fix it

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+
+def err_auto_result_name_on_non_def_decl : Error<
+   "postcondition with result name %0 cannot appear on a function declaration with deduced return type unless it is a definition">;
+   // "result name '%0' cannot appear in postcondition for a non-template function declaration with a deduced return type">;
```
I think the commented version is better, I would remove the "non-template" bit

cor3ntin commented on clang/include/clang/Basic/DiagnosticSemaKinds.td:
```diff
@@ -12542,6 +12549,128 @@ def warn_coroutine_type_aware_allocator_ignored : Warning <
+def err_auto_result_name_on_non_def_decl : Error<
+   "postcondition with result name %0 cannot appear on a function declaration with deduced return type unless it is a definition">;
+   // "result name '%0' cannot appear in postcondition for a non-template function declaration with a deduced return type">;
+def note_function_return_type : Note<
+    "function return type declared here">;
```
Is that useful?

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+  DeclarationInfo.Contracts = Seq;
+}
+
+StmtResult Parser::ParseFunctionContractSpecifierImpl(
+    llvm::function_ref<QualType()> ReturnTypeResolver, ContractScopeOffset ScopeOffset, bool &IsInvalid) {
```
Can we 
 -  call that `ParseContractAfterKeyword` or something
 -  Pass in parameter the kind (pre/post/assert) and the kw location?

it would let us not mix the handling of the assert kw and that of pre/post - they appear in different contexts and work differently

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+
+static float EOFData = 0.0;
+
+ContractSpecifierDecl *
+Parser::ParseLexedFunctionContractsInScope(CachedTokens &ContractToks,
```
Is that used?

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+    if (Contract.isUsable())
+      Contracts.push_back(Contract.getAs<ContractStmt>());
+  }
+  ContractSpecifierDecl *Seq = Actions.ActOnFinishContractSpecifierSequence(
+      Contracts, StartLoc, IsInvalid);
+
```
You should check we are parsing a function (and not a type or a function pointer)
see `err_requires_clause_on_declarator_not_declaring_a_function`

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+      Actions, Sema::ExpressionEvaluationContext::PotentiallyEvaluated);
+
+  ResultNameDecl *RND = nullptr;
+  // FIXME(EricWF): We allow parsing the result name declarator in `pre` so we
+  // can diagnose it but we don't do the same for contract assert... Should we?
+  if ((CK != ContractKind::Assert) && Tok.is(tok::identifier) &&
```
We probably should **not**!

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+
+    ExprLoc = ConsumeToken();
+    QualType ReturnType;
+    if (ReturnTypeResolver) {
+      ReturnType = ReturnTypeResolver();
+    }
```
```suggestion
    if (ReturnTypeResolver)
      ReturnType = ReturnTypeResolver();
```

just because i noticed :)

cor3ntin commented on clang/lib/Parse/ParseContracts.cpp:
```diff
@@ -0,0 +1,437 @@
+
+  T.consumeClose();
+
+  if (Cond.isInvalid()) {
+    Cond =
+        Actions.CreateRecoveryExpr(ExprLoc, EndLoc, {}, Actions.Context.BoolTy);
+  } else {
+    SetInvalidOnExit.release();
+  }
+
```
Same here for the braces

EricWF committed:
merge with main

EricWF committed:
merge with main

EricWF committed:
merge changes

EricWF committed:
remove temp changes

EricWF committed:
Merge branch 'contracts-nightly' of github.com:efcs/llvm-project into contracts-nightly

EricWF committed:
Merge contracts-base into contracts-nightly

Resolved merge conflicts by keeping both sides of changes:
- clang/include/clang/AST/Decl.h: Added both ContractKind enum and UsualDeleteParams struct
- clang/lib/AST/DeclPrinter.cpp: Added both VisitResultNameDecl and VisitTemplateTemplateParmDecl functions
- clang/lib/Frontend/CompilerInvocation.cpp: Added both contract group serialization and AllocToken code

EricWF committed:
Merge contracts-base into contracts-nightly

Resolved merge conflicts:
- clang/lib/Basic/IdentifierTable.cpp: Kept enum definitions with KEYCONTRACTS support
- libcxx/utils/libcxx/test/features.py: Kept modified version (modify-delete conflict)

EricWF committed:
Fix build errors after merge

After merging contracts-base into contracts-nightly, several build errors
needed to be addressed:

1. TokenKey and KeywordStatus enums were moved to IdentifierTable.h in
   contracts-base, but contracts-nightly had added KEYCONTRACTS to the
   .cpp file version. Added KEYCONTRACTS to the header and removed the
   duplicate enum definitions from the .cpp file.

2. ArrayRef no longer accepts std::nullopt for construction in newer LLVM.
   Updated return statements to use {} instead of std::nullopt in:
   - FunctionDecl::contracts() in Decl.cpp
   - Sema::getInterveningContractScopes() in SemaContract.cpp

EricWF committed:
add constification flag to be off

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
actually remove injected type declarations

EricWF committed:
remove debug things

EricWF committed:
fix missing assert file

EricWF committed:
add new debug

EricWF committed:
temp debugging

EricWF committed:
fix various crashes

EricWF committed:
Update clang/include/clang/AST/Decl.h

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

EricWF committed:
Update clang/lib/Parse/ParseContracts.cpp

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

EricWF committed:
remove unused function

EricWF committed:
Update clang/include/clang/AST/ASTContext.h

Co-authored-by: Corentin Jabot <corentinjabot@gmail.com>

EricWF committed:
remove unused variable

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
add flag  to disable lambda capture restrictions

cor3ntin commented on clang/tools/libclang/CXCursor.cpp:
```diff
@@ -662,6 +662,11 @@ CXCursor cxcursor::MakeCXCursor(const Stmt *S, const Decl *Parent,
     K = CXCursor_CXXParenListInitExpr;
     break;
 
+  // FIXME(EricWF): Add support for the following Stmt classes.
```
needs not be in the initial pr
```suggestion
  // FIXME: Add support for the following Stmt classes.
```

cor3ntin commented on clang/lib/Serialization/ASTReaderStmt.cpp:
```diff
@@ -509,6 +509,22 @@ void ASTStmtReader::VisitDependentCoawaitExpr(DependentCoawaitExpr *E) {
 
+void ASTStmtReader::VisitContractStmt(ContractStmt *S) {
+  VisitStmt(S);
+  Record.skipInts(1);
```
this needs a comment

cor3ntin commented on clang/lib/Serialization/ASTReaderStmt.cpp:
```diff
@@ -4449,6 +4465,17 @@ Stmt *ASTReader::ReadStmtFromStream(ModuleFile &F) {
 
+    case STMT_CXX_CONTRACT: {
+      BitsUnpacker ContractBits(Record[ASTStmtReader::NumStmtFields]);
+      ContractKind CK = static_cast<ContractKind>(ContractBits.getNextBits(2));
```
we might want `2` to be a constant in ContractsOptions

cor3ntin commented on clang/lib/Serialization/ASTReaderDecl.cpp:
```diff
@@ -4154,6 +4192,13 @@ Decl *ASTReader::ReadDeclRecord(GlobalDeclID ID) {
+    break;
+  case DECL_CONTRACT_SPECIFIER:
+    D = ContractSpecifierDecl::CreateDeserialized(Context, ID,
+                                                  Record.readInt());
```
```suggestion
                                                  /*NumContracts=*/Record.readInt());
```

cor3ntin commented on clang/lib/Serialization/ASTCommon.cpp:
```diff
@@ -399,6 +399,10 @@ bool serialization::isRedeclarableDeclKind(unsigned Kind) {
+  // FIXME(EricWF): Is this correct? I think so?
+  case Decl::ContractSpecifier:
+    return true;
+
```
I need to understand why it is Decl to begin with

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -15871,11 +15936,15 @@ TreeTransform<Derived>::TransformLambdaExpr(LambdaExpr *E) {
   }
 
+  bool WasInContract =
+      getSema().currentEvaluationContext().isContractAssertionContext();
```
That could be move in `PushExpressionEvaluationContextForFunction`

cor3ntin commented on clang/lib/Sema/TreeTransform.h:
```diff
@@ -8941,6 +8955,50 @@ TreeTransform<Derived>::TransformCoyieldExpr(CoyieldExpr *E) {
+  EnterExpressionEvaluationContext Unevaluated(
+      SemaRef, Sema::ExpressionEvaluationContext::PotentiallyEvaluated, nullptr,
+      Sema::ExpressionEvaluationContextRecord::EK_Other,
+      /*ShouldEnter=*/S->getContractKind() != ContractKind::Assert);
```
`Unevaluated` is exactly the wrong name :)

EricWF committed:
Finish merge

EricWF committed:
fix silly build error

EricWF committed:
Merge branch 'contracts-base' into contracts-nightly

EricWF committed:
fix build error after removal af _LIBCPP_WEAK
